### ISSUE TITLE
Multi word searches or "quoted search" 

### ISSUE THREAD

**Issue by hlodver** [barseghyanartur/django-elasticsearch-dsl-drf#20]
## Original issue (opened by @hlodver)

How can I do a search like Kibana does with quotes and people are used to use with google?
`/api/v1/search/?search="Day%20trips"`

I could not find anything in the documentation

## Update (added by @barseghyanartur)

Implement the following search filter backends (read the thread from top to the bottom for discussions).

- [ ] [match_phrase](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query-phrase.html)
- [ ] [match_phrase_prefix](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query-phrase-prefix.html)
- [x] [multi_match](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#CO190-1)
- [ ] [common_terms](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-common-terms-query.html)
- [ ] [query_string](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html)
- [x] [simple_query_string](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html)

**Comment by barseghyanartur** [comment]
@hlodver:

Elastic ``operator`` options for are currently not supported. 

I'm adding it to the TODO list. 

I can't give a certain indication about when would it "arrive" in the package though.

**Comment by barseghyanartur** [comment]
For the reference:

https://www.elastic.co/guide/en/elasticsearch/guide/current/match-multi-word.html

**Comment by barseghyanartur** [comment]
@mrchrisadams:

Look, Elastic has a lot of options for that.

All listed [here](https://www.elastic.co/guide/en/elasticsearch/reference/current/full-text-queries.html).

You're probably most interested in the following:

- [match_phrase](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-match-query-phrase.html)
- [multi_match](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#CO190-1)
- [query_string](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-query-string-query.html)
- [simple_query_string](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html)

I must confess, that I didn't have much time to spend on the search filter backend, however, it needs major improvements in order to support all these things. I think finally all of them need to be implemented (as separate backends).

I'll try to start with [simple_query_string](https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-simple-query-string-query.html).

Contributions are greatly appreciated. If you're willing to spend some time on this, we can discuss how to divide the work.

**Comment by barseghyanartur** [comment]
@mrchrisadams:

FYI, before query_string and simple_query_string are implemented, there's a need to find an alternative value separator [here](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/issues/63).

**Comment by mrchrisadams** [comment]
Oh sweet, thanks @barseghyanartur - the `simple_query_string` looks like it does much more than what I need, so I don't mind having a go at one of these.

It looks like `match_phrase` only works on a single field, so I don't think it would work so well with how you specific the search fields in the Viewset, but `multi_match` _does_ allow for phrase matching, across multiple fields.

https://www.elastic.co/guide/en/elasticsearch/reference/current/query-dsl-multi-match-query.html#type-phrase


From what I can see the different filters work by having their own reserved param, like `?search` in the SearchFilter case, or in the case of the filter, listening for all the non reserved params, then checking for varous filters to add like `__gt` or `__contains` and so on.

I'm wondering if having an extra `?phrase_search` param that carries out a phrase search on the fields provided might work, and support the same modifiers as the current search.

So you might do 

```
search/?phrase_search=Day%20trips
```

Which would end up as this kind of query:

```json
{
  "query": {
    "multi_match": {
        "query": "Day Trips",
        "type":"phrase",
        "fields": [ "all_of_the", "search_fields", "defined_in_the_viewset"]
    }
  }
}
```

To search all the normal search fields,

```
search/?phrase_search=specific_field|Day%20trips
```

Would end up with:

```json
{
  "query": {
    "multi_match": {
        "query": "Day Trips",
        "type":"phrase",
        "fields": [ "all_of_the", "search_fields", "defined_in_the_viewset"]
    }
  }
}
```

I haven't looked into how elasticsearch-dsl treats MultiMatch queries yet, or whether it lets you mix them with other search or filter querie.  If this is possible, I'll update here.

Also, I'm not fussy about the final name, `phrase_search` is just for these examples here.

Anyway, thanks for the response. I've switched on github notifcations again, so I'll actually get emails when you respond now…

**Comment by barseghyanartur** [comment]
@mrchrisadams:

Working branch for this is [search_backends](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/tree/search-backends).

The first steps are done. An old single search filter backend (`SearchFilterBackend`) is to be replaced with new style search filter backends (at the moment, the `django_elasticsearch_dsl_drf.filter_backends.search.compound.CompoundSearchFilterBackend` resembles all the functionality of the `SearchFilterBackend`).

Differences:

(1) Pluggable query backends.

```python
class CompoundSearchFilterBackend(BaseSearchFilterBackend):
    """Compound search backend."""

    query_backends = [
        MatchQueryBackend,
        NestedQueryBackend,
    ]
```

(2) Configurabe `matching` options. Default is `should`, but can be easily changed to `must` in your custom search query backend.

You can pretty much base your code on this functionality, as it will be genrally available soon. Tests are passing.

You could for instance, declare match_all query backend and plug it into. I will myself work on that further, but any help is welcome.

**Comment by mrchrisadams** [comment]
Awesome, thanks for @barseghyanartur - I'll look into this tomorrow.

Are you proposing `match_all` query backend be something like a backend that has `must` as the default, instead of the `should` in the current one?

**Comment by barseghyanartur** [comment]
@mrchrisadams:

- `match_all` as a Query Backend: Yes.
- `must` instead of `should`: No, since it's not necessary. Instead, you can configure it per Search Filter Backend.

**Comment by mrchrisadams** [comment]
wow @barseghyanartur this is really cool! I had a go at setting up a test harness yesterday, with the vagrant box, but I think I had something misconfigured, as I couldn't get tests to pass locally on my laptop.

I'm going to have another go today to see if I can add some tests for the multi match stuff, and if I can't get them running, I'll file an issue explaining what I had trouble with.

Thanks again for this 👍

**Comment by barseghyanartur** [comment]
I don't recommend using Elastic 2.x.
Better use Elastic 5.x or 6.x. There's a doc here install_elasticsearch.rst (in the root). There are docker containers ready to use.

**Comment by mrchrisadams** [comment]
Okay, @barseghyanartur, I've got the test suite running here with `python runtests.py` and elastic search 5.x now. 

Thanks for the pointers.

So, to set up a test for this, I'm assuming I'd create a separate TestMultiSearch class, and in the setup, add the new [`BookMultiMatchSearchFilterBackendDocumentViewSet`][2] view, and [`MultiMatchSearchFilterBackend`][3]

[2]: https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/search-backends/examples/simple/search_indexes/viewsets/book.py#L427

[3]: https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/search-backends/src/django_elasticsearch_dsl_drf/filter_backends/search/multi_match.py#L27

I can largely follow what's going on in the [`_search_by_field`][4] method in your test class, so I'm assuming I'd so something  similar, but with `search_multi_match` as the search param.

[4]: https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/search-backends/src/django_elasticsearch_dsl_drf/tests/test_search.py#L101

Can you let me know if this is about right? If so, I'll have a look at it tonight

**Comment by barseghyanartur** [comment]
@mrchrisadams:

Yep. You've got it all right!

**Comment by barseghyanartur** [comment]
Once there are tests for what's implemented at the moment, I'll make a new release (0.12). Thus, with introduction of new-style Search Filter Backends, Multi Match Search Filter Backend and Compound Search Filter Backend (as well as a couple of other intermediate fixes).

Implementation of ``match_phrase``, ``query_string`` and ``simple_query_string`` backends is moved to 0.13 milestone.

**Comment by barseghyanartur** [comment]
0.12 is released.

**Comment by barseghyanartur** [comment]
The ``simple_query_string`` implementation landed in the [search_backends](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/tree/search-backends) branch. Tests are passing. [Documentation](https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/search-backends/search_backends.rst) available. Comments are welcome.

**Comment by barseghyanartur** [comment]
``simple_query_string`` search filter backend is implemented in 0.13.

The issue is considered to be solved. Implementation of ``match_phrase``, ``match_phrase_prefix``,  ``common_terms`` and ``query_string`` backends shall still be implemented, however, I hope that someone (not me) would pick the task.
